smart_simulation/ds_tools/sma_forecast.py

```diff
@@ -0,0 +1,303 @@
+        start_weight, consumption_series_test
+    )
+    test_theoretical_binary = test_weight_binary(start_weight, consumption_series_test)
+    pred_consumption = predict(sma_window, consumption_series_train, test_dates)
+    pred_weight = calculate_test_weights(start_weight, pred_consumption)
+    pred_binary = test_weight_binary(start_weight, pred_consumption)
+
+    data_dict["train_weight"] = training_weights
+    data_dict["train_consumption"] = consumption_series_train
+    data_dict["test_weight"] = test_theoretical_weight
+    data_dict["test_consumption"] = consumption_series_test
+    data_dict["test_binary"] = test_theoretical_binary
+    data_dict["pred_weight"] = pred_weight
+    data_dict["pred_consumption"] = pred_consumption
+    data_dict["pred_binary"] = pred_binary
+
+    return data_dict
+
+
+def multi_test(
+    weight_series: pd.Series, train_test_dict: dict, sma_window: int
+) -> dict:
+    """
+    Forecast consumption for a all train/test datasets in train_test_dict.
+    Args:
+        weight_series: Full weight series associated with the train_test_dict.
+        train_test_dict: Dictionary of train/test datasets.
+        sma_window: Size of the averaging window to use for forecasting.
+
+    Returns: A dictionary of sma_single_test dictionaries.
+    """
+    all_tests_dict = {}
+    pred_dates = list(train_test_dict.keys())
+    for day in pred_dates:
+        day_train_test = train_test_dict[day]
+        consumption_series_train = day_train_test["train"]
+        consumption_series_test = day_train_test["test"]
+        weights_series_train = weight_series[consumption_series_train.index].copy()
+        day_pred_dict = single_test(
+            weights_series_train,
+            consumption_series_train,
+            consumption_series_test,
+            sma_window,
+        )
+        all_tests_dict[day] = day_pred_dict
+
+    return all_tests_dict
+
+
+def create_test_df(test_dict: dict, prediction_date: pd.Timestamp) -> pd.DataFrame:
+    """
+    Create a Pandas DataFrame of test data from test_dict. Drop training data before returning DataFrame.
+    Args:
+        prediction_date: Date of prediction.
+        test_dict: Test dict of forecast results.
+    Returns: Test data DataFrame.
+    """
+    copy_dict = test_dict.copy()
+    del copy_dict["train_weight"]
+    del copy_dict["train_consumption"]
+    test_df = pd.DataFrame.from_dict(copy_dict, orient="columns")
+    test_df.loc[:, "date_of_prediction"] = prediction_date
+    test_df.index.name = "date"
+    test_df = test_df.reset_index()
+    return test_df
+
+
+def create_all_tests_df(multi_test_dict: dict) -> pd.DataFrame:
+    """
+    Create a Pandas DataFrame of test data from all test_dicts in a multi_test_dict.
+    Args:
+        multi_test_dict: Dictionary of test_dicts for all dates with predictions.
+    Returns: Test data DataFrame.
+    """
+    test_dfs = {}
+    pred_dates = list(multi_test_dict.keys())
```

Comment:
You don't need to gather the keys into a list before using it in the `for` loop. If you use it directly it will iterate over the keys: 
``` for day in multi_test_dict:```

Reply:
And since you're also using the value right below you can:
```
for day, test in multi_test_dict.items():
```

Reply:
Making this change now. The code is much cleaner this way, thanks.